server/src/main/java/org/apache/druid/server/http/MetadataResource.java

```diff
@@ -349,37 +353,52 @@ public Response getUnusedSegmentsInDataSource(
-    }
-    if (limit != null && limit < 0) {
-      throw InvalidInput.exception("Invalid limit[%s] specified. Limit must be > 0", limit);
-    }
+    try {
+      if (dataSource == null || dataSource.isEmpty()) {
+        throw InvalidInput.exception("dataSourceName must be non-empty.");
+      }
 
-    if (lastSegmentId != null && SegmentId.tryParse(dataSource, lastSegmentId) == null) {
-      throw InvalidInput.exception("Invalid lastSegmentId[%s] specified.", lastSegmentId);
-    }
+      if (limit != null && limit < 0) {
+        throw InvalidInput.exception("Invalid limit[%s] specified. Limit must be > 0.", limit);
+      }
 
-    SortOrder theSortOrder = sortOrder == null ? null : SortOrder.fromValue(sortOrder);
+      if (lastSegmentId != null && SegmentId.tryParse(dataSource, lastSegmentId) == null) {
+        throw InvalidInput.exception("Invalid lastSegmentId[%s] specified.", lastSegmentId);
+      }
 
-    final Interval theInterval = interval != null ? Intervals.of(interval.replace('_', '/')) : null;
-    Iterable<DataSegmentPlus> unusedSegments = segmentsMetadataManager.iterateAllUnusedSegmentsForDatasource(
-        dataSource,
-        theInterval,
-        limit,
-        lastSegmentId,
-        theSortOrder
-    );
+      final SortOrder theSortOrder = sortOrder == null ? null : SortOrder.fromValue(sortOrder);
+
+      final Interval theInterval = interval != null ? Intervals.of(interval.replace('_', '/')) : null;
+      final Iterable<DataSegmentPlus> unusedSegments = segmentsMetadataManager.iterateAllUnusedSegmentsForDatasource(
+          dataSource,
+          theInterval,
+          limit,
+          lastSegmentId,
+          theSortOrder
+      );
 
-    final Function<DataSegmentPlus, Iterable<ResourceAction>> raGenerator = segment -> Collections.singletonList(
-        AuthorizationUtils.DATASOURCE_READ_RA_GENERATOR.apply(segment.getDataSegment().getDataSource()));
+      final Function<DataSegmentPlus, Iterable<ResourceAction>> raGenerator = segment -> Collections.singletonList(
+          AuthorizationUtils.DATASOURCE_READ_RA_GENERATOR.apply(segment.getDataSegment().getDataSource()));
 
-    final Iterable<DataSegmentPlus> authorizedSegments =
-        AuthorizationUtils.filterAuthorizedResources(req, unusedSegments, raGenerator, authorizerMapper);
+      final Iterable<DataSegmentPlus> authorizedSegments =
+          AuthorizationUtils.filterAuthorizedResources(req, unusedSegments, raGenerator, authorizerMapper);
 
-    final List<DataSegmentPlus> retVal = new ArrayList<>();
-    authorizedSegments.iterator().forEachRemaining(retVal::add);
-    return Response.status(Response.Status.OK).entity(retVal).build();
+      final List<DataSegmentPlus> retVal = new ArrayList<>();
+      authorizedSegments.iterator().forEachRemaining(retVal::add);
+      return Response.status(Response.Status.OK).entity(retVal).build();
+    }
+    catch (DruidException e) {
```

Comment:
Could we not instead map `InvalidInput` exception to the appropriate response in some higher layer exception mapper so that all apis benefit, instead of doing this try catch here?